Rewatch.lua

```diff
@@ -307,24 +312,58 @@ function rewatch_SetTooltip(data)
 	end;
 end;
 
+-- function rewatch_CreateSpell(spellId)
+-- 	local completeSpell;
+-- 	local spell = Spell:CreateFromSpellID(spellId);
+-- 	return spell;
+-- end
+
+-- gets the name of a spell
+-- spellId: the spell ID of the spell to get its name
+-- return: the corresponding spellName 
+function rewatch_GetSpellName(spellId)
+	local spell = Spell:CreateFromSpellID(spellId);
+	local spellName = spell:GetSpellName();
+	return spellName;
+end
+
 -- gets the spell ID of the highest rank of the specified spell
 -- spellName: the name of the spell to get the highest ranked spellId from
 -- return: the corresponding spellId
 function rewatch_GetSpellId(spellName)
 	-- get spell info and highest rank, return if the user can't cast it (not learned, etc)
-	local name, rank, icon = GetSpellInfo(spellName);
+	local name, rank, icon, _, _, _, spellId = GetSpellInfo(spellName);
 	if(name == nil) then return -1; end;
-	-- loop through all book spells, return the number if it matches above data
-	local i, ispell, irank = 1, GetSpellBookItemName(1, BOOKTYPE_SPELL);
-	repeat
-		if ((ispell == name) and ((rank == irank) or (not irank))) then return i; end;
-		i, ispell, irank = i+1, GetSpellBookItemName(i+1, BOOKTYPE_SPELL);
-	until (not ispell);
-	
-	-- return default -1
-	return -1;
+
+	local ispellId = -1
+	local spell = Spell:CreateFromSpellID(spellId);
+	--spell:ContinueOnSpellLoad(function() ispellId = rewatch_GetSpellIdFromSpellBook(spellId, name, rank); end);
+	spell:ContinueOnSpellLoad(function(spellId, name, rank) --ispellId =
+		local i, ispell, irank = 1, GetSpellBookItemName(1, BOOKTYPE_SPELL);
+		repeat
+			if ((ispell == name) and ((rank == irank) or (not irank) or (rank == nil or irank == nil))) then return i; end;
+			i, ispell, irank = i+1, GetSpellBookItemName(i+1, BOOKTYPE_SPELL);
+		until (not ispell);
+		
+		-- return default -1
+		return -1;
+	end);
+
+	return ispellId;
```

Comment:
Blizzard have made some major back end changes with the Spell System API, with a focus on SpellId. The spell text fields, except SpellName are now loaded on demand.
This means that 'If you call these functions, the listed return value may be nil or empty, since the data is not available at the time of the function call.'

GetSpellBookItemName -- # 2 (spell name subtext) is affected by this change. (I believe this is retrieved by irank).

I have made the above changes folling the Blizzard examples.

See [ 8.0 Release Notes for Wow Addon Developers - Blue post](https://us.battle.net/forums/en/wow/topic/20762318007#post-1)

Reply:
Apologies , I wasn't sure how many lines it would include. I refer to lines 337-352.